MRG: Preliminary stage of rewrite (#36)
resampling-stats/resampling-with: source/resampling_with_code.Rmd

Put another way, `some_number` is now the name we can use for this collection
of 10 values.

{{< var array_name >}}s are very useful for simulations and data analysis, and
we will be using for nearly every example in this book.

## Functions

We can use [`rnd.choice`]{.python}[the `sample` function]{.r} to select a single
value *at random* from the sequence of numbers in `some_numbers`.

```{python, opts.label="py_ed"}
# Select a number from the choices in some_numbers
my_number = rnd.choice(some_numbers)
# Show the value that results.
my_number
```

```{r, opts.label="r_ed"}
# Select a number from the choices in some_numbers
my_number <- sample(some_numbers)
# Show the value that results.
my_number
```

[`rnd.choice`]{.python}[the `sample` function]{.r} is a *function*.

:::python
:::{.callout-note}
## Functions and methods

Actually, to be precise, we should call `rnd.choice` a *method*.  A method is a
*function attached to a value*. In this case the function `choice` is attached
to the value `rnd`.  That's not an important distinction for us at the moment,
so please forgive our strategic imprecision, and let us continue to say that
`rnd.choice` is a function.
:::
:::

> We will go into more detail on functions later.  For now, think of a function as a *production line*.  We sent the function raw material to work on.  The production line does some work on the raw materials.  A finished result comes off the other end.

Author: I've dropped the recipe analogy now.

a > 0
```

There are 17 values in `a`, so the comparison to 0 means there are 17 answers.
{{< var np_or_r >}} therefore returns a[n]{.python} *{{< var array_name >}}*
of 17 elements, containing these 17 answers.  The first answer is the answer to
the question "is the value of the first element of `a` greater than 0", and the
second is the answer to "is the value of the second element of `a` greater than
0".

Let us store the result of this comparison to work on:

```{python, opts.label="py_ed"}
# Is the value of "a" greater than 0
# Store as another array "q".
q = a > 0
# Show the value of q
q
```

```{r, opts.label="r_ed"}
# Is the value of "a" greater than 0?
# Store as another array "q".
q <- a > 0
# Show the value of q
q
```

Notice that there is one
[`True`]{.python}[`TRUE`]{.r}
element in `q` for every element in `a` that was greater than 0.  It only
remains to count the number of
[`True`]{.python}[`TRUE`]{.r}
values in `q`, to get the count of patients in our simulated trial who were
cured.

::: python
NumPy has a function that will do that count of `True` elements, called
`np.count_nonzero`.  It will count the number of elements in an array that are

> not equivalent to zero.  `False` is equivalent to 0 in Python:

Author: Rewritten.

introduces our first *statistical* problem, where we use probability to draw
conclusions about some important data — about a potential cure for a type of
cancer.  We will not make much of the distinction between probability and
statistics here, but we will come back to it several times in later chapters.

<!---
*TODO* - Check we have discussed this before.
-->

## A new treatment for Burkitt lymphoma

In 2006, researchers in the US National Cancer Institute tested a new
treatment for Burkitt lymphoma.  [Burkitt
lymphoma](https://en.wikipedia.org/wiki/Burkitt_lymphoma) is a rather unusual
cancer of the lymphatic system — a vein-like network throughout the body that
is involved in the immune reaction to disease. In developed countries, with
standard treatment, the cure rate for Burkitt lymphoma is about 90%.  The
National Cancer Institute's gave their new treatment to 17 patients, and found
that all 17 patients were "cured" — meaning that the disease had not
progressed in any of the 17 patients, in more than two years of follow-up.

Here is where we put on our statistical hat.  Now we are wearing that hat, we
wonder whether the 100% could be due to random variation.

> How could it happen by chance, that 17 out of 17 patients were cured?

Reviewer: Or maybe "Could it happen, by chance, that ..."?

> We know that the average cure rate in the developed world is 90%.  That means,
> that each patient that we treat has a 90% chance of being cured.  Let's say we
> treat one patient with some new treatment, and they are cured.   Is the
> treatment better or worse than the standard treatment?   We don't really know,
> because there is a 90% chance they would be cured, if the treatment is just
> the same as the standard treatment, and so it is more likely than not that
> they will be cured, even if the treatment as about the same as the standard
> treatment.

Reviewer: ```suggestion
We know that the average cure rate in the developed world is 90%.  Lacking any further information, we could therefore think of each patient being treated as having a 90% chance of being cured.  If a patient now receives the new treatment and is cured, does this tell us whether the new treatment is better than the standard treatment? We don't really know: as long as the new treatment isn't significantly worse than the standard treatment, curing the patient is the expected outcome.
```

special that warrants further investigation, or is it something that could reasonably have come about by chance?

This is exactly the kind of question we ask in statistics — is this result
*surprising* or *unusual*, given that we know that there is an element of
chance?

First we come up with a *model* of the world we expect, when the treatment is
90% effective.  This model has to allow for *chance* because each patient has
a 90% *chance* of cure.

::: python
To put chance into our model, we will be asking the computer to generate many
random numbers.  So, before we start, we again import NumPy and get its
*random number generator*:

```{python, opts.label="py_ed"}
import numpy as np

# Ask for NumPy's default random number generator and name
# it `rnd`.  `rnd` is short for "random".
rnd = np.random.default_rng()
```
:::

It is often useful to think how we would make a physical model of chance in
this world.  For example, to simulate whether a given patient is cured or not
by a 90% effective treatment, we could throw a ten sided die and record the
result.  We could say, arbitrarily, that a result of 0 means "not cured", and

> all the numbers 1 through 9 mean "cured" (typical 10-sided dice have sides

Reviewer: 10-sided dice are likely labeled 1 through 10, like the 6-sided versions

Reviewer: https://rpg.stackexchange.com/a/73939

It is often useful to think how we would make a physical model of chance in
this world.  For example, to simulate whether a given patient is cured or not
by a 90% effective treatment, we could throw a ten sided die and record the
result.  We could say, arbitrarily, that a result of 0 means "not cured", and
all the numbers 1 through 9 mean "cured" (typical 10-sided dice have sides
numbered 0 through 9).

We could roll 17 dice to simulate one "trial" in this random world.  For each
trial, we record the number of dice that show numbers 1 through 9 (and not
0).  This will be a number between 0 and 17, and it is the number of patients
"cured" in our simulated trial.

Here is the result of one such trial we did with a set of 17 10-sided dice we happened to have near to hand:

:::{.callout-note}
We need a photo of 17 10-sided dice here
:::

In this case there is one 0, and the rest are from 1 through 9, so there were 16 out of 17 "patients" who were "cured" in this trial.

<!---
Adapt above to picture.
-->

We can use the computer to do something very similar, by asking the computer
for 17 random whole numbers from 0 through 9.

:::{.callout-note}
## Whole numbers

> A whole number is a number that is not negative, and does not have fractional

Reviewer: Whole numbers are a woolly concept.  Maybe rather go with something explicit like integers from 0 through 9.

Author: OK to resolve this one?

part (does not have anything after a decimal point).  0 and 1 and 2 and 3 are
whole numbers, but -1 and $\frac{3/5}$ and 11.3 are not. The whole numbers
from 0 through 9 are 0, 1, 2, 3, 4, 5, 6, 7, 8, 9.
:::

:::{.callout-note}
## What do we mean by *random*?

We have been throwing around the term random, but now we will say a little
more about what we mean by it.

> We say something is random when we cannot completely predict the result.  The randomness is the part we cannot predict.

Reviewer: You explain this below in a way that makes sense to me.  I wonder, however, if people won't get hung up on this statement.  Surely the weather isn't random even though we cannot predict it!

WIth a coin toss, I guess one could say that it is fully deterministic given knowledge of starting conditions (hand position, velocity of flip, air resistance, etc.).  But for all practical purposes it is random.

So, there is that distinction (that you make below) between inherent randomness and practical randomness.

Let us say we toss a coin, and we could get either a *head* or a *tail*.  Let
us also say that we know for sure that 50% of the time we get a head and 50%
we get a tail.  We can predict some things about this coin toss.  We can
predict that it will either be a head or a tail, and we can predict that it
will be a head 50% of the time. But we can't predict *which one* of heads or
tails we will get on the next throw.  We say the result of each throw is

> *random* if we cannot do anything to improve our prediction of 50% for heads

Reviewer: This rings true to me: there's nothing we can do to improve our guess.

(or tails) on the next toss.

When we ask for a random whole number from 0 through 9, we mean that we can't
do anything to predict which of the 10 possible numbers we will get on any one
trial.  We can't do better than guessing that the next value will be — say — 0
— and whichever number we guess, we will only ever have a 10% chance of being
correct.

The computer uses lots of tricks to give us numbers that are random, in a
practical sense — meaning that there is no *practical* way for you or anyone
else to predict the next number.  These practical-sense random numbers are

> also called *pseudo-random* numbers.  The pseudo-random numbers from

Reviewer: :+1:

> [Numpy]{.python}[R]{.r} use some very serious machinery to make the numbers

Reviewer: ```suggestion
[NumPy]{.python}[R]{.r} use some very serious machinery to make the numbers
```

hard to predict, and practically random.  No-one can guarantee that it is
absolutely *impossible* to find some way of improving our prediction for the
next pseudo-random number that

> [Numpy]{.python}[R]{.r}

Reviewer: ```suggestion
[NumPy]{.python}[R]{.r}
```

gives us, but we, and the rest of the scientific world, do not know of any way
to do it. That practical sense of randomness from the computer is good enough
for every case that we cover in this book, and we bet it will also be true for

> every possible case that you will come across.

Reviewer: ```suggestion
every case that you will come across in practice.
```

maybe?

Here is the result of one such trial we did with a set of 17 10-sided dice we happened to have near to hand:

![One roll of 17 10-sided dice](images/17_d10s.png)

In this case there are four 0s, and the rest are from 1 through 9, so there
were 13 out of 17 not-zero numbers, and therefore, 13 out of 17 "patients" who
were "cured" in this simulated trial.

<!---
Adapt text above to picture.
-->

We could repeat this simulated trial procedure 100 times, and we would then
have 100 counts of the not-zero numbers. Each of the 100 counts would be the
number of patients cured in that trial.  Then we could see what proportion of
the simulated trials had 17 out of 17 patients cured.   If the 17 out of 17
count is only a small proportion — say, less than 5% — then we conclude that,

> if 17 Burkitt Lymphoma patients have the standard treatment, it is *unlikely*

Reviewer: This sentence gets a bit long.

# it `rnd`.  `rnd` is short for "random".
rnd = np.random.default_rng()
```

> We prepare the list of numbers that we want Numpy to select from:

Reviewer: ```suggestion
We prepare the list of numbers that we want NumPy to select from:
```

```{python, opts.label="py_ed"}
# Make the list of numbers, store with the name "numbers".

> numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Reviewer: I thought we're working with 0-9, because of the dice labeling.
